Implement Watcher APIs

russcam opened the pull request:
This is a rather large PR 😄 I would like to highlight some areas that would be great to get some feedback on

1. The watcher stats API is the first rest api spec that defines a URL path parameter as an enum. I've accommodated this as a special case but wonder if there is some more general approach that can be taken here
2. The inputs to the `PutWatch` API are generally concrete implementations of `*Container` types to constrain the input to accepting a single value out of a set of available values; the `PutWatchApiTests` class object initializer example is a good example to look at e.g. `ConditionContainer`, `TriggerContainer`, etc. this follows the design we have for `QueryContainer` et. al. The `GetWatch` API however exposes the `Watch` results as the interfaces, as accessing the values of a watch through the interface seems more natural than casting a concrete type that explicitly implements it.
3. The `ArrayCompareConditionDescriptor` seems very unwieldy in taking 4 parameters in a method call. I'm in two minds about whether introducing a descriptor here would make the implementation cleaner.
4. The `ExecutionResultAction` `*ActionResult` seem like they would be a good candidate for generalizing an ActionResult type. Thoughts?

gmarz commented on src/Nest/XPack/Watcher/Condition/CompareConditionBase.cs:
```diff
@@ -0,0 +1,151 @@
+
+	public class CompareConditionDescriptor : IDescriptor
+	{
+		// TODO: Should this follow DescriptorBase<TDescriptor, TInterface>?
```
I'm fine with leaving this as is, especially if both `path` and `value` are always required.

russcam replied:
ok, removed comment

gmarz commented on src/Nest/XPack/Watcher/Condition/InlineScriptCondition.cs:
```diff
@@ -0,0 +1,35 @@
+	}
+
+	public class InlineScriptConditionDescriptor :
+	ScriptConditionDescriptorBase<InlineScriptConditionDescriptor, IInlineScriptCondition>, IInlineScriptCondition
```
indentation?

russcam commented on src/Nest/CommonAbstractions/Infer/Id/Id.cs:
```diff
@@ -5,7 +5,7 @@
 {
 	[JsonConverter(typeof(IdJsonConverter))]
-	public class Id : IUrlParameter
+	public class Id : IEquatable<Id>, IUrlParameter
```
How do you feel about deserializing ids from Elasticsearch into `Id` instances?

russcam commented on src/Nest/CommonAbstractions/CoordinatedRequestObserverBase.cs:
```diff
@@ -1,14 +1,15 @@
 		private readonly Action<T> _onNext;
 		private readonly Action<Exception> _onError;
-		private readonly Action _completed;
+		private readonly System.Action _completed;
```
fixed

russcam commented on src/Nest/CommonAbstractions/Response/ApiCallDetailsOverride.cs:
```diff
@@ -17,7 +17,7 @@ internal class ApiCallDetailsOverride : IApiCallDetails
 
 		public Exception OriginalException => this._original.OriginalException;
-		public HttpMethod HttpMethod => this._original.HttpMethod;
+		public Elasticsearch.Net.HttpMethod HttpMethod => this._original.HttpMethod;
```
fixed

russcam commented on src/Nest/CommonAbstractions/SerializationBehavior/GenericJsonConverters/EpochDateTimeJsonConverter.cs:
```diff
@@ -0,0 +1,53 @@
+
+namespace Nest
+{
+	internal class EpochDateTimeJsonConverter : DateTimeConverterBase
```
done

russcam commented on src/Nest/Modules/SnapshotAndRestore/Restore/RestoreObservable/RestoreObserver.cs:
```diff
@@ -5,12 +5,12 @@ namespace Nest
+			Action<IRecoveryStatusResponse> onNext = null,
 			Action<Exception> onError = null,
-			Action completed = null)
+			System.Action completed = null)
```
fixed

russcam commented on src/Nest/CommonAbstractions/SerializationBehavior/GenericJsonConverters/ReadSingleOrEnumerableJsonConverter.cs:
```diff
@@ -0,0 +1,45 @@
+		}
+	}
+
+	// TODO: Evaluate if this is needed
```
removed

russcam commented on src/Nest/Modules/SnapshotAndRestore/Snapshot/SnapshotObservable/SnapshotObservable.cs:
```diff
@@ -63,7 +63,7 @@ public IDisposable Subscribe(IObserver<ISnapshotStatusResponse> observer)
 				_snapshotStatusHumbleObject.Error += onError;
 
-				_timer = new Timer(Snapshot, observer, _interval, Timeout.InfiniteTimeSpan);
+				_timer = new Timer(Snapshot, observer, _interval, System.Threading.Timeout.InfiniteTimeSpan);
```
fixed

russcam commented on src/Nest/Modules/SnapshotAndRestore/Snapshot/SnapshotObservable/SnapshotObserver.cs:
```diff
@@ -5,12 +5,12 @@ namespace Nest
+			Action<ISnapshotStatusResponse> onNext = null,
 			Action<Exception> onError = null,
-			Action completed = null)
+			System.Action completed = null)
```
fixed

russcam commented on src/Nest/Search/Count/CountRequest.cs:
```diff
@@ -19,17 +19,17 @@ public partial interface ICountRequest<T> : ICountRequest
 		private CountRequestParameters QueryString => ((IRequest<CountRequestParameters>)this).RequestParameters;
-		protected override HttpMethod HttpMethod =>
-			this.QueryString.ContainsKey("_source") || this.QueryString.ContainsKey("q") ? HttpMethod.GET : HttpMethod.POST;
+		protected override Elasticsearch.Net.HttpMethod HttpMethod =>
```
fixed

russcam commented on src/Tests/CodeStandards/Descriptors.doc.cs:
```diff
@@ -49,16 +75,7 @@ where isGeneric
 				where isFunc
                 let firstFuncArg = type.GetGenericArguments().First()
```
fixed

russcam commented on src/Tests/Framework/EndpointTests/ApiTestBase.cs:
```diff
@@ -43,7 +43,7 @@ public abstract class ApiTestBase<TCluster, TResponse, TInterface, TDescriptor,
 
 		protected abstract string UrlPath { get; }
-		protected abstract HttpMethod HttpMethod { get; }
+		protected abstract Elasticsearch.Net.HttpMethod HttpMethod { get; }
```
fixed

russcam commented on src/Tests/Framework/ManagedElasticsearch/FileSystem/TestRunnerFileSystem.cs:
```diff
@@ -177,9 +178,67 @@ public void BeforeStart(IEnumerable<string> settings)
+					"xpack.notification.slack:",
+					"  account:",
+					"    monitoring:",
+					"      url: https://hooks.slack.com/services/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",
```
fixed. I've made it a dummy endpoint as it just needs to have some value

russcam commented on src/Tests/Framework/VirtualClustering/Audit/Auditor.cs:
```diff
@@ -69,7 +69,7 @@ public void ChangeTime(Func<DateTime, DateTime> selector)
 			this.AssertPoolBeforeCall?.Invoke(this._cluster.ConnectionPool);
 
-			Action call = () => this._cluster.ClientCall(callTrace?.RequestOverrides);
+			System.Action call = () => this._cluster.ClientCall(callTrace?.RequestOverrides);
```
fixed

russcam commented on src/Nest/XPack/Watcher/ActivateWatch/ActivateWatchRequest.cs:
```diff
@@ -0,0 +1,15 @@
+﻿namespace Nest
+{
+	public partial interface IActivateWatchRequest
+	{
```
fixed

russcam commented on src/Nest/XPack/Watcher/Condition/ArrayCompareConditionBase.cs:
```diff
@@ -0,0 +1,211 @@
+
+			while (reader.Read())
+			{
+				if (reader.TokenType == JsonToken.PropertyName)
```
changed 👍

russcam commented on src/Nest/XPack/Watcher/Condition/ArrayCompareConditionBase.cs:
```diff
@@ -0,0 +1,211 @@
+
+						while (reader.Read())
+						{
+							if (reader.TokenType == JsonToken.PropertyName)
```
changed 👍

russcam commented on src/CodeGeneration/ApiGenerator/Domain/ApiEndpoint.cs:
```diff
@@ -265,12 +265,38 @@ public IEnumerable<CsharpMethod> CsharpMethods
+				var metric = this.Url.Parts.First(p => p.Key == "metric");
+
+				var apiUrlPart = metric.Value;
+				apiUrlPart.Name = "watcher_stats_metric";
```
The renaming is really for our code generation and enum naming purposes, but if it would help others, I will.

russcam commented:
I've addressed the review comments in https://github.com/elastic/elasticsearch-net/pull/2391/commits/91d1ab762f05925e3b381a4a0a5064fc5e4b890b.

If the changes look good, I'll rebase, squash and merge

Mpdreamz commented on src/Nest/CommonAbstractions/Union/UnionJsonConverter.cs:
```diff
@@ -12,8 +12,6 @@ internal class UnionJsonConverter : JsonConverter
 		private static readonly ConcurrentDictionary<Type, UnionJsonConverterBase> KnownTypes = new ConcurrentDictionary<Type, UnionJsonConverterBase>();
 
 		public override bool CanConvert(Type objectType) => true;
-		public override bool CanRead => true;
```
while not needed I like the explicit self documenting implementations of implementing `CanRead` and `CanWrite`

gmarz commented:
LGTM!

russcam commented:
ported to `master` in 14bd45d0990f85896908cb39431b549bcf82f2f4

Will keep this open until port to `2.x` is done.
